src/cuter_cerl.erl

```diff
@@ -163,47 +169,85 @@
-      Exports = extract_exports(M, AST),
-      {TypeAttrs, SpecAttrs} = classify_attributes(cerl:module_attrs(AST)),
-      Types = cuter_types:retrieve_types(TypeAttrs),
-      Specs = cuter_types:retrieve_specs(SpecAttrs),
-      Defs = cerl:module_defs(AST),
-      Funs = [process_fundef(D, Exports, M, TagGen) || D <- Defs],
-      {ok, kmodule(M, Types, Specs, Funs)};
+      {ok, kmodule(M, AST, TagGen)};
     {error, _} = Error -> Error
   end.
 
 %% -------------------------------------------------------------------
 %% kmodule API
 %% -------------------------------------------------------------------
 
--spec kmodule(module(), cuter_types:stored_types(), cuter_types:stored_specs(), [{mfa(), kfun()}]) -> kmodule().
-kmodule(M, Types, Specs, Funs) ->
+%% Constructs a kmodule.
+%% It takes the AST of the module as input and performs all the computations
+%% to produce a kmodule.
+-spec kmodule(module(), cerl:cerl(), tag_generator()) -> kmodule().
+kmodule(M, AST, TagGen) ->
   Kmodule = ets:new(M, [ordered_set, protected]),
+  Exports = extract_exports(M, AST),
+  Attrs = cerl:module_attrs(AST),
+  RecordForms = extract_record_forms(Attrs),
+  TypeForms = extract_type_forms(Attrs),
+  SpecForms = extract_spec_forms(Attrs),
+  Types = cuter_types:retrieve_types(TypeForms, RecordForms),
+  Specs = cuter_types:retrieve_specs(SpecForms),
+  Defs = cerl:module_defs(AST),
+  Funs = [process_fundef(D, Exports, M, TagGen) || D <- Defs],
+  ets:insert(Kmodule, {name, M}),
   ets:insert(Kmodule, {types, Types}),
   ets:insert(Kmodule, {specs, Specs}),
+  ets:insert(Kmodule, {spec_forms, SpecForms}),
+  ets:insert(Kmodule, {exported_types, extract_exported_types(M, Attrs)}),
+  ets:insert(Kmodule, {type_forms, TypeForms}),
+  ets:insert(Kmodule, {record_forms, RecordForms}),
   lists:foreach(fun({Mfa, Kfun}) -> ets:insert(Kmodule, {Mfa, Kfun}) end, Funs),
   Kmodule.
 
+%% Returns a set of all the exported types of a kmodule.
+%% Types are module-prefixed.
+-spec kmodule_exported_types(kmodule()) -> sets:set({module(), atom(), arity()}).
+kmodule_exported_types(Kmodule) ->
+  [{exported_types, ExpTypes}] = ets:lookup(Kmodule, exported_types),
+  ExpTypes.
+
+%% Returns the name of a kmodule.
+-spec kmodule_name(kmodule()) -> module().
+kmodule_name(Kmodule) ->
+  [{name, Name}] = ets:lookup(Kmodule, name),
+  Name.
+
+%% Returns the processed specs of a kmodule.
 -spec kmodule_specs(kmodule()) -> cuter_types:stored_specs().
 kmodule_specs(Kmodule) ->
   [{specs, Specs}] = ets:lookup(Kmodule, specs),
   Specs.
 
+%% Returns the processed types of a kmodule.
 -spec kmodule_types(kmodule()) -> cuter_types:stored_types().
 kmodule_types(Kmodule) ->
   [{types, Types}] = ets:lookup(Kmodule, types),
   Types.
 
-%% Retrieves the kfun() for the given MFA.
+%% Returns the unprocessed types (and opaques) of a kmodule (as forms).
+-spec kmodule_type_forms(kmodule()) -> [{integer(), cerl:cerl()}].
```

Comment:
It's better to give a (type) name to this `integer()` so that it's clear what it denotes.